### PR TITLE
Migrating to node 16 to fix audit

### PR DESCRIPTION
**Pull request by andrewrlee**
### Context

https://trello.com/c/hdMgIvup

### Intent

* Miigrate to node 16
* Miigrate from node-sass to dart sass (as node-sass no longer maintained)
* Remove annoying `class-methods-use-this` and `max-classes-per-file` eslint rules

### Checklist

- [ ] This PR contains **only** changes related to the above card
- [ ] Tests have been added/updated to cover the change
- [ ] Documentation has been updated where appropriate
- [ ] Tested in Development

### CODE REVIEW

File: bin/build-css [hunk 1]
```diff
@@ -1,5 +1,9 @@
 #!/bin/bash
 
-./node_modules/node-sass/bin/node-sass $@ \
-    assets/sass/style.scss \
-    assets/stylesheets/application.css
+./node_modules/.bin/sass \
+    --no-source-map \
+    --quiet-deps \
```
Comment (andrewrlee) [inline bin/build-css:5]: we get some deprecation warnings with dart-sass due to [this defect](https://github.com/alphagov/govuk-frontend/issues/2238).

The --quiet-deps flag drops it from many to 3 warnings

File: package.json [hunk 2]
```diff
@@ -56,33 +55,33 @@
     "moment": "^2.29.1",
     "morgan": "^1.10.0",
     "nocache": "^3.0.1",
-    "node-sass-middleware": "0.11.0",
```
Comment (andrewrlee) [inline package.json:59]: no longer have prod dependency on node-sass-middleware and node-sass which often have lots of vulnerabilities associated with them.

File: server/auth/middleware.js [hunk 3]
```diff
@@ -34,7 +34,7 @@ const _authenticate = (req, res, next) =>
   });
 
 function isPrisonerId(id) {
-  const pattern = new RegExp(/^[A-Z][0-9]{4}[A-Z]{2}$/i);
```
Comment (andrewrlee) [inline server/auth/middleware.js:37]: new eslint rule detects unnecessary wrapping of regexs

File: server/services/offender/responses/importantDates.js [hunk 4]
```diff
@@ -1,15 +1,8 @@
-const { parseISO, format, isValid } = require('date-fns');
 const {
   placeholders: { DEFAULT },
   dateFormats: { PRETTY_DATE },
 } = require('../../../utils/enums');
-
-const formatDateOrDefault = (defaultValue = '', dateFormat, date) => {
```
Comment (andrewrlee) [inline server/services/offender/responses/importantDates.js:7]: this function was duplicated now pulling from elsewhere

File: server/utils/__tests__/date.spec.js [hunk 5]
```diff
@@ -65,28 +65,6 @@ describe('DateUtils', () => {
 });
 
 describe('getDateSelectionFrom', () => {
-  it('defaults to generating the previous 12 months from today', () => {
```
Comment (andrewrlee) [inline server/utils/__tests__/date.spec.js:68]: this defaulting functionality is never used and eslint now picks up with only having the first param as a default

File: audit-ci.json [hunk 6]
```diff
@@ -3,5 +3,5 @@
   "moderate": true,
   "package-manager": "auto",
   "registry": "https://registry.npmjs.org",
-  "allowlist": [1002401, 1002624, 1002503, 1002546]
+  "allowlist": []
```
Comment (andrewrlee) [inline audit-ci.json:6]: no longer required to be allowed

File: Dockerfile [hunk 7]
```diff
@@ -12,15 +12,17 @@ WORKDIR /app
 
 COPY . .
 
-RUN npm ci --no-audit --production && \
+RUN npm ci --no-audit --ignore-scripts && \
```
Comment (andrewrlee) [inline Dockerfile:15]: don't run husky

File: Dockerfile [hunk 8]
```diff
@@ -12,15 +12,17 @@ WORKDIR /app
...
     npm run build && \
     export BUILD_NUMBER=${BUILD_NUMBER:-1_0_0} && \
     export GIT_REF=${GIT_REF:-dummy} && \
     export GIT_DATE="${GIT_DATE:-dummy}" && \
-    npm run record-build-info
+    npm run record-build-info && \
+    npm prune --production
```
Comment (andrewrlee) [inline Dockerfile:21]: remove dev dependencies, once we're done with them

**Review by jonhindle — APPROVED** [review]
lgtm

**Pull request merged**